c/rc_lock.c

```diff
@@ -0,0 +1,464 @@
+    if (err == CKB_INDEX_OUT_OF_BOUND) {
+      return false;
+    }
+    CHECK(err);
+
+    if (memcmp(lock_script_hash, buff, BLAKE160_SIZE) == 0) {
+      return true;
+    }
+    i += 1;
+  }
+
+exit:
+  return false;
+}
+
+int verify_identity(IdentityType *id, SmtProofEntryVecType *proofs,
+                    RceState *rce_state) {
+  int err = 0;
+  uint32_t proof_len = proofs->t->len(proofs);
+  CHECK2(proof_len == rce_state->rcrules_count, ERROR_PROOF_LENGTH_MISMATCHED);
+
+  uint8_t key[SMT_KEY_BYTES] = {0};
+  key[0] = id->flags;
+  memcpy(key + 1, id->blake160, BLAKE160_SIZE);
+
+  smt_pair_t entries[1];
+  smt_state_t states;
+  smt_state_init(&states, entries, 1);
+  smt_state_insert(&states, key, SMT_VALUE_EMPTY);
+
+  uint8_t proof_mask = 0x3;  // both
+  for (uint32_t i = 0; i < proof_len; i++) {
+    bool existing = false;
+    SmtProofEntryType proof_entry = proofs->t->get(proofs, i, &existing);
+    CHECK2(existing, ERROR_INVALID_MOL_FORMAT);
+    mol2_cursor_t proof = proof_entry.t->proof(&proof_entry);
+
+    const RCRule *current_rule = &rce_state->rcrules[i];
+    err = rce_verify_one_rule(rce_state, &states, NULL, NULL, proof_mask, proof,
+                              current_rule);
+    CHECK(err);
+  }
+  if (rce_state->has_wl) {
+    if (rce_state->both_on_wl) {
+      err = 0;
+    } else {
+      err = ERROR_NOT_ON_WHITE_LIST;
+    }
+  } else {
+    // all black list, it's not allowed
+    err = ERROR_NO_WHITE_LIST;
+  }
+exit:
+  return err;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main() {
+#endif
+  // parse lock script's args
+  int err = 0;
+  ArgsType args = {0};
+
+  err = parse_args(&args);
+  CHECK(err);
+
+  // extra signature from witness
+  WitnessArgsType witness;
+  err = make_witness(&witness);
+  CHECK(err);
+
+  BytesOptType lock = witness.t->lock(&witness);
+  CHECK2(!lock.t->is_none(&lock), ERROR_INVALID_MOL_FORMAT);
+  mol2_cursor_t lock_bytes = lock.t->unwrap(&lock);
+  // convert Bytes to RcLockWitnessLock
+  RcLockWitnessLockType witness_lock = make_RcLockWitnessLock(&lock_bytes);
+
+  // pubkey hash
+  if (args.id.flags == IdentityFlagsPubkeyHash ||
+      args.id.flags == IdentityFlagsPubkeyHashRc) {
+    uint8_t signature_bytes[SIGNATURE_SIZE];
+    BytesOptType signature_opt = witness_lock.t->signature(&witness_lock);
+    mol2_cursor_t signature_cursor = signature_opt.t->unwrap(&signature_opt);
+
+    uint32_t read_len =
+        mol2_read_at(&signature_cursor, signature_bytes, SIGNATURE_SIZE);
+    CHECK2(read_len == SIGNATURE_SIZE, ERROR_INVALID_MOL_FORMAT);
+
+    err = verify_secp256k1_blake160_sighash_all(args.id.blake160,
+                                                signature_bytes);
+    CHECK(err);
+  }
+
+  // owner lock
+  if (args.id.flags == IdentityFlagsOwnerLock ||
+      args.id.flags == IdentityFlagsOwnerLockRc) {
+    bool existing = is_lock_script_hash_existing(args.id.blake160);
```

Comment:
This is incorrect, in case the flags is `IdentityFlagsOwnerLockRc`, the actual lock script hash should be provided in the witness, and validated by SMT